File: DSCI553/hw2/task1.py
import sys
import time
from pyspark import SparkContext
from itertools import combinations

def apriori(baskets, support):
    basket_list = list(baskets)
    if not basket_list:
        return []
    
    item_counts = {}
    for basket in basket_list:
        for item in basket:
            item_counts[item] = item_counts.get(item, 0) + 1
    
    frequent = {}
    k = 1
    frequent[k] = set()
    for item, count in item_counts.items():
        if count >= support:
            frequent[k].add((item,))
    
    while frequent[k]:
        k += 1
        candidates = set()
        freq_list = list(frequent[k-1])
        for i in range(len(freq_list)):
            for j in range(i+1, len(freq_list)):
                union = tuple(sorted(set(freq_list[i]) | set(freq_list[j])))
                if len(union) == k:
                    valid = True
                    for subset in combinations(union, k-1):
                        if subset not in frequent[k-1]:
                            valid = False
                            break
                    if valid:
                        candidates.add(union)
        
        candidate_counts = {}
        for basket in basket_list:
            basket_set = set(basket)
            for candidate in candidates:
                if set(candidate).issubset(basket_set):
                    candidate_counts[candidate] = candidate_counts.get(candidate, 0) + 1
    
        frequent[k] = set()
        for candidate, count in candidate_counts.items():
            if count >= support:
                frequent[k].add(candidate)
        
        if not frequent[k]:
            break
    
    result = []
    for k in sorted(frequent.keys()):
        for itemset in frequent[k]:
            result.append(itemset)
    
    return result


if __name__ == '__main__':

    case_num = int(sys.argv[1])
    support = int(sys.argv[2])
    input_file = sys.argv[3]
    output_file = sys.argv[4]
    
    start_time = time.time()

    sc = SparkContext(appName='task1')
    sc.setLogLevel('ERROR')
    
    data = sc.textFile(input_file)
    header = data.first()
    data = data.filter(lambda x: x != header)
    
    pairs = data.map(lambda x: x.split(','))
    
    if case_num == 1:
        baskets = pairs.map(lambda x: (x[0], x[1])) \
                       .groupByKey() \
                       .map(lambda x: (x[0], set(x[1]))) \
                       .map(lambda x: x[1])
    else:
        baskets = pairs.map(lambda x: (x[1], x[0])) \
                       .groupByKey() \
                       .map(lambda x: (x[0], set(x[1]))) \
                       .map(lambda x: x[1])
    
    num_partitions = baskets.getNumPartitions()
    
    partition_support = support / num_partitions
    
    candidates = baskets.mapPartitions(lambda part: apriori(part, partition_support)) \
                        .distinct() \
                        .collect()
    
    candidates_by_size = {}
    for candidate in candidates:
        size = len(candidate)
        if size not in candidates_by_size:
            candidates_by_size[size] = []
        candidates_by_size[size].append(candidate)
    
    for size in candidates_by_size:
        candidates_by_size[size].sort()
    
    candidates_set = set(candidates)
    
    def count_candidates(baskets):
        basket_list = list(baskets)
        counts = {}
        for basket in basket_list:
            basket_set = set(basket)
            for candidate in candidates_set:
                if set(candidate).issubset(basket_set):
                    counts[candidate] = counts.get(candidate, 0) + 1
        return counts.items()
    
    candidate_counts = baskets.mapPartitions(count_candidates) \
                              .reduceByKey(lambda a, b: a + b) \
                              .filter(lambda x: x[1] >= support) \
                              .map(lambda x: x[0]) \
                              .collect()
    
    frequent_by_size = {}
    for itemset in candidate_counts:
        size = len(itemset)
        if size not in frequent_by_size:
            frequent_by_size[size] = []
        frequent_by_size[size].append(itemset)
    
    for size in frequent_by_size:
        frequent_by_size[size].sort()
    
    with open(output_file, 'w') as f:
        f.write("Candidates:\n")
        for size in sorted(candidates_by_size.keys()):
            items = candidates_by_size[size]
            if size == 1:
                result = ','.join([f"('{item[0]}')" for item in items])
            else:
                result = ','.join([str(item) for item in items])
            f.write(result + "\n\n")
        
        f.write("Frequent Itemsets:\n")
        for size in sorted(frequent_by_size.keys()):
            items = frequent_by_size[size]
            if size == 1:
                result = ','.join([f"('{item[0]}')" for item in items])
            else:
                result = ','.join([str(item) for item in items])
            f.write(result + "\n\n")
    
    sc.stop()
    
    end_time = time.time()
    duration = int(end_time - start_time)
    print(f"Duration: {duration}")

File: DSCI553/hw2/hw2_project/src/main/scala/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import scala.collection.mutable

object task1 {
  
  implicit val listOrdering: Ordering[List[String]] = Ordering.by((_: List[String]).mkString(","))
  
  def apriori(baskets: Iterator[Set[String]], support: Double): Iterator[List[String]] = {
    val basketList = baskets.toList
    if (basketList.isEmpty) {
      return Iterator.empty
    }
    
    val itemCounts = mutable.Map[String, Int]()
    for (basket <- basketList; item <- basket) {
      itemCounts(item) = itemCounts.getOrElse(item, 0) + 1
    }
    
    val frequentSingles = itemCounts.filter(_._2 >= support).keys.toSet
    var resultItemsets = mutable.ListBuffer[List[String]]()
    
    for (item <- frequentSingles.toList.sorted) {
      resultItemsets += List(item)
    }
    
    if (frequentSingles.size < 2) {
      return resultItemsets.iterator
    }
    
    val sortedSingles = frequentSingles.toList.sorted
    var candidatePairs = mutable.Map[List[String], Int]()
    
    for (i <- sortedSingles.indices; j <- i + 1 until sortedSingles.length) {
      candidatePairs(List(sortedSingles(i), sortedSingles(j))) = 0
    }
    
    for (basket <- basketList; pair <- candidatePairs.keys) {
      if (pair.forall(basket.contains)) {
        candidatePairs(pair) += 1
      }
    }
    
    val frequentPairs = candidatePairs.filter(_._2 >= support).keys.toSet
    resultItemsets ++= frequentPairs.toList.sorted(listOrdering)
    
    if (frequentPairs.isEmpty) {
      return resultItemsets.iterator
    }
    
    val filteredBaskets = basketList.map(basket => 
      basket.filter(frequentSingles.contains).toList
    ).filter(_.length >= 3)
    
    var currentFrequent = frequentPairs
    var k = 3
    
    while (currentFrequent.nonEmpty) {
      val allItems = currentFrequent.flatten.toSet
      val candidatesK = allItems.toList.sorted.combinations(k).toList
      
      if (candidatesK.isEmpty) {
        return resultItemsets.iterator
      }
      
      var candidateCounts = mutable.Map[List[String], Int]()
      for (candidate <- candidatesK) {
        candidateCounts(candidate) = 0
      }
      
      for (basket <- filteredBaskets; candidate <- candidatesK) {
        if (candidate.forall(basket.contains)) {
          candidateCounts(candidate) += 1
        }
      }
      
      val nextFrequent = candidateCounts.filter(_._2 >= support).keys.toSet
      if (nextFrequent.isEmpty) {
        return resultItemsets.iterator
      }
      
      resultItemsets ++= nextFrequent.toList.sorted
      currentFrequent = nextFrequent
      k += 1
    }
    
    resultItemsets.iterator
  }
  
  def countCandidates(baskets: Iterator[Set[String]], candidates: Set[List[String]]): Iterator[(List[String], Int)] = {
    val basketList = baskets.toList
    val counts = mutable.Map[List[String], Int]()
    
    for (basket <- basketList; candidate <- candidates) {
      if (candidate.forall(basket.contains)) {
        counts(candidate) = counts.getOrElse(candidate, 0) + 1
      }
    }
    
    counts.iterator
  }
  
  def formatOutput(itemsets: List[List[String]]): String = {
    val grouped = itemsets.groupBy(_.length).toList.sortBy(_._1)
    val lines = mutable.ListBuffer[String]()
    
    for ((size, items) <- grouped) {
      val sortedItems = items.sorted
      val formatted = if (size == 1) {
        sortedItems.map(item => s"('${item.head}')").mkString(",")
      } else {
        sortedItems.map(item => s"(${item.map(x => s"'$x'").mkString(",")})").mkString(",")
      }
      lines += formatted
    }
    
    lines.mkString("\n\n")
  }
  
  def main(args: Array[String]): Unit = {
    val startTime = System.currentTimeMillis()
    
    val caseNum = args(0).toInt
    val support = args(1).toInt
    val inputFile = args(2)
    val outputFile = args(3)
    
    val conf = new SparkConf().setAppName("task1")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")
    
    val data = sc.textFile(inputFile)
    val header = data.first()
    val pairs = data.filter(_ != header).map(_.split(","))
    
    val baskets = if (caseNum == 1) {
      pairs.map(arr => (arr(0), arr(1)))
           .groupByKey()
           .map { case (_, businesses) => businesses.toSet }
    } else {
      pairs.map(arr => (arr(1), arr(0)))
           .groupByKey()
           .map { case (_, users) => users.toSet }
    }
    
    baskets.cache()
    
    val numPartitions = baskets.getNumPartitions
    val partitionSupport = support.toDouble / numPartitions
    
    val candidates = baskets.mapPartitions(part => apriori(part, partitionSupport))
                           .distinct()
                           .collect()
                           .toList
    
    val candidatesOutput = formatOutput(candidates)
    
    val candidateSet = candidates.toSet
    val candidateBroadcast = sc.broadcast(candidateSet)
    
    val frequent = baskets.mapPartitions(part => countCandidates(part, candidateBroadcast.value))
                         .reduceByKey(_ + _)
                         .filter(_._2 >= support)
                         .keys
                         .collect()
                         .toList
    
    val frequentOutput = formatOutput(frequent)
    
    val writer = new java.io.PrintWriter(outputFile)
    try {
      writer.write("Candidates:\n")
      writer.write(candidatesOutput)
      writer.write("\n\n")
      writer.write("Frequent Itemsets:\n")
      writer.write(frequentOutput)
    } finally {
      writer.close()
    }
    
    baskets.unpersist()
    sc.stop()
    
    val duration = (System.currentTimeMillis() - startTime) / 1000
    println(s"Duration: $duration")
  }
}